src/dom/Struct.ts

```diff
@@ -0,0 +1,96 @@
+import {Value, PathElement} from "./Value";
+import {IonTypes} from "../Ion";
+import {DomValue} from "./DomValue";
+
+/**
+ * Represents a struct[1] value in an Ion stream.
+ *
+ * [1] http://amzn.github.io/ion-docs/docs/spec.html#struct
+ */
+export class Struct extends DomValue(Object, IonTypes.STRUCT) implements Value {
+    /**
+     * Constructor.
+     * @param fields        An iterator of field name/value pairs to represent as a struct.
+     * @param annotations   An optional array of strings to associate with this null value.
+     */
+    constructor(fields: IterableIterator<[string, Value]>, annotations: string[] = []) {
+        super();
+        for(let [fieldName, value] of fields) {
+            Object.defineProperty(this, ''+fieldName, {
+                configurable: true,
+                enumerable: true,
+                value: value,
+                writable: true
+            })
+        }
+        this._setAnnotations(annotations);
+    }
```

Comment:
I think this is burying the lede if I am reading this correctly.  You do surface field names as properties, yes?  If so, this should be documented explicitly in the doc comment.

What happens when they incidentally shadow methods or the internal fields?  Does this need to gate against that?

Reply:
> You do surface field names as properties, yes?

Yes.

> If so, this should be documented explicitly in the doc comment.

Agreed, I'll add this.

> What happens when they incidentally shadow methods or the internal fields? Does this need to gate against that?

My plan (not implemented in this already quite large PR) is for all property accesses on structs to resolve to method names and internal fields first. If you have data that would shadow one of those fields, it can be accessed by `.get(fieldName)` instead. The iterators returned by `fields()` and `fieldNames()` would behave properly, returning the custom property values instead of the methods/fields.

```typescript
let s = ion.load('{stringValue: "Hello", foo: "bar"}')!; // The `stringValue` field shadows Value#stringValue()
let greeting = s.stringValue; // This resolves to the method handle and is probably not what the user wanted
let greeting = s['stringValue']; // This resolves to the method handle and is probably not what the user wanted
let greeting = s.stringValue(); // `greeting` is now "Hello"
let greeting = s.get('stringValue'); // `greeting is now "Hello"
```

This means that there are ~16 `Struct` field names that can have unexpected behavior, which seemed a better compromise than simply forbidding them.